### PR TITLE
Implemented Dark Mode for Android (#25525)

### PR DESCRIPTION
**Pull request by matthew-carroll**
Starting in Android Pie, when battery saver is on, or when the developer option for "Night Mode" is enabled, Android apps should show a dark styled UI. This PR introduces this behavior to Flutter.

This is the framework half of the change.  The engine PR here is here:
https://github.com/flutter/engine/pull/7488

On the framework side, this PR integrates the new `platformBrightness` property into MediaQuery and then binds WidgetsApp to that property so that the widget hierarchy rebuilds on change to brightness.

Supporting Flutter changes for brightness required adding uiMode to the Android manifest under overridden configuration changes for FlutterActivity.

### CODE REVIEW

File: packages/flutter/lib/src/material/app.dart [hunk 1]
```diff
@@ -166,6 +169,17 @@ class MaterialApp extends StatefulWidget {
   /// The colors to use for the application's widgets.
   final ThemeData theme;
 
+  /// The colors to use for the application's widgets when the host platform has
+  /// specifically requested a dark UI.
+  ///
+  /// For example, starting with Android Pie, turning on battery saver causes
```
Comment (xster) [inline packages/flutter/lib/src/material/app.dart]: You can use a stronger tone here. e.g. this is only implemented on Android and only on Android Pie (I thought you said Q) and only when androidmanifest says X and your android view has attribute Y.

File: packages/flutter/lib/src/material/app.dart [hunk 2]
```diff
@@ -166,6 +169,17 @@ class MaterialApp extends StatefulWidget {
...
+  /// apps to go into "dark mode".
+  ///
```
Comment (xster) [inline packages/flutter/lib/src/material/app.dart]: Directly address how this is different from (Cupertino)ThemeData.brightness

File: packages/flutter/lib/src/widgets/app.dart [hunk 3]
```diff
@@ -1021,6 +1021,15 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
     });
   }
 
+  // RENDERING
+  @override
+  void didChangePlatformBrightness() {
+    setState(() {
+      // The platformBrightness property of ui.window has changed. We reference
```
Comment (xster) [inline packages/flutter/lib/src/widgets/app.dart]: I think this will work but it's a bit too coincidental. The actual references and usages are in MaterialApp which won't rebuild.

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 4]
```diff
@@ -109,6 +111,15 @@ class MediaQueryData {
...
   ///    textScaleFactor defined for a [BuildContext].
   final double textScaleFactor;
+  
+  /// The current brightness mode of the host platform.
+  /// 
+  /// For example, starting in Android Pie, battery saver mode asks all apps to
+  /// render in a "dark mode". That change is reflected in this property.
+  /// 
+  /// Not all platforms necessarily support a concept of brightness mode. Those
+  /// platforms will report [PlatformBrightness.no_preference] in this property.
```
Comment (xster) [inline packages/flutter/lib/src/widgets/media_query.dart]: Other enums in dart:ui are camelCased

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 5]
```diff
@@ -523,6 +540,12 @@ class MediaQuery extends InheritedWidget {
     return MediaQuery.of(context, nullOk: true)?.textScaleFactor ?? 1.0;
   }
 
+  /// Returns platformBrightness for the nearest MediaQuery ancestor or
+  /// [PlatformBrightness.no_preference], if no such ancestor exists.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: It would be worth pointing out that context will be rebuilt if any aspect of the MediaQuery changes

File: packages/flutter/lib/src/widgets/app.dart (hunk 3 above)
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/app.dart]: This shouldn't be necessary at all. MaterialApp's theme Builder should rebuild when the brightness changes because MediaQuery.of()

File: packages/flutter/lib/src/material/app.dart [hunk 6]
```diff
@@ -403,45 +417,63 @@ class _MaterialAppState extends State<MaterialApp> {
...
+            MaterialPageRoute<T>(settings: settings, builder: builder),
+      home: widget.home,
+      routes: widget.routes,
+      initialRoute: widget.initialRoute,
+      onGenerateRoute: widget.onGenerateRoute,
+      onUnknownRoute: widget.onUnknownRoute,
+      builder: (BuildContext context, Widget child) {
+        // Use a light theme, dark theme, or fallback theme.
+        ThemeData theme;
+        final ui.PlatformBrightness platformBrightness = MediaQuery.platformBrightnessOf(context);
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: This should cause the Builder widget to rebuild when any change to to the MediaQuery occurs.

File: packages/flutter/lib/src/material/app.dart (hunk 1 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: @xster I provided a present-day example, but I avoided using language that limits the example because it is a near certainty that this behavior will be available in new contexts and across multiple platforms.  If we claim that it is only for battery saver on Android Pie then we're just creating a future task to reword that, right? (assuming someone remembers to reword it)

File: packages/flutter/lib/src/material/app.dart (hunk 2 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: @xster can you describe how the 2 would be confused?  I'm not sure how to explain the difference because I'm not sure how they're similar to begin with....

File: packages/flutter/lib/src/widgets/app.dart (hunk 3 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/app.dart]: I removed the setState() as per @HansMuller's comment

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 4 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/media_query.dart]: I'm in the process of trying to remove this enum and replace it with `Brightness`

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 5 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/media_query.dart]: Done

File: packages/flutter/lib/src/material/app.dart (hunk 6 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: @HansMuller is there a change you wanted me to make here?

File: packages/flutter/lib/src/material/app.dart (hunk 2 above)
Comment (xster) [reply packages/flutter/lib/src/material/app.dart]: One is an incoming signal from the platform and one is an outgoing effect that describes what descendent widgets look like.

File: packages/flutter/lib/src/material/app.dart [hunk 7]
```diff
@@ -166,6 +169,17 @@ class MaterialApp extends StatefulWidget {
   /// The colors to use for the application's widgets.
   final ThemeData theme;
 
+  /// The colors to use for the application's widgets when the host platform has
+  /// specifically requested a dark UI.
+  ///
+  /// For example, starting with Android Pie, turning on battery saver causes
+  /// apps to go into "dark mode".
+  ///
+  /// See also
```
Comment (xster) [inline packages/flutter/lib/src/material/app.dart]: Be consistent with the rest of the codebase

```
/// See also:
///
/// * blah
```

This form might render to `See also - blah` in some markdown engines.

File: packages/flutter/lib/src/services/system_chrome.dart [hunk 8]
```diff
@@ -419,4 +406,4 @@ class SystemChrome {
   @visibleForTesting
   static SystemUiOverlayStyle get latestStyle => _latestStyle;
   static SystemUiOverlayStyle _latestStyle;
-}
+}
```
Comment (xster) [inline packages/flutter/lib/src/services/system_chrome.dart]: revert this line

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 9]
```diff
@@ -59,6 +58,7 @@ class MediaQueryData {
     this.size = Size.zero,
     this.devicePixelRatio = 1.0,
     this.textScaleFactor = 1.0,
+    this.platformBrightness = ui.Brightness.light,
     this.padding = EdgeInsets.zero,
     this.viewInsets = EdgeInsets.zero,
     this.alwaysUse24HourFormat = false,
```
Comment (xster) [inline packages/flutter/lib/src/widgets/media_query.dart]: @goderbauer, while you're here :D

I left a #27311. Null values for these fields don't really make much sense.

**Review by xster — COMMENTED** [review]
LG. Please add tests

File: packages/flutter/lib/src/material/app.dart (hunk 7 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Replaced "-" with "*"

File: packages/flutter/lib/src/services/system_chrome.dart (hunk 8 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/services/system_chrome.dart]: Done

File: packages/flutter/lib/src/material/app.dart [hunk 10]
```diff
@@ -166,6 +169,17 @@ class MaterialApp extends StatefulWidget {
   /// The colors to use for the application's widgets.
```
Comment (xster) [inline packages/flutter/lib/src/material/app.dart]: Here's a proposed doc. 

```dart
/// The [ThemeData] to use when the platform's requested brightness is [Brightness.light] or undefined. 
/// 
/// Host platforms such as Android Pie can request a system-wide "dark mode" when entering battery saver mode.
/// 
/// When the host platform requests a [Brightness.light] mode, you may want to supply a [ThemeData.brightness]
/// that's also [Brightness.light].
///
/// Defaults to a light [ThemeData.fallback] theme when null.
///
/// See also:
///
/// * [MediaQueryData.platformBrightness], which indicates the platform's desired brightness and is used to
///    automatically toggle between [theme] and [darkTheme] in [MaterialApp].
/// * [ThemeData.brightness], which indicates the [Brightness] of a [ThemeData] you may want to use as a 
///    consequence of the platform requesting a light or dark mode.
final ThemeData theme;

/// The [ThemeData] to use when the platform specifically requests a dark themed UI.
///
/// Host platforms such as Android Pie can request a system-wide "dark mode" when entering battery saver mode.
/// 
/// When the host platform requests a [Brightness.dark] mode, you may want to supply a [ThemeData.brightness]
/// that's also [Brightness.dark].
///
/// Uses [theme] instead when null. Defaults to a light [ThemeData.fallback] theme when both [darkTheme] and
/// [theme] are null.
///
/// See also:
///
/// * [MediaQueryData.platformBrightness], which indicates the platform's desired brightness and is used to
///    automatically toggle between [theme] and [darkTheme] in [MaterialApp].
/// * [ThemeData.brightness], which indicates the [Brightness] of a [ThemeData] you may want to use as a 
///    consequence of the platform requesting a light or dark mode.
final ThemeData darkTheme;

**Comment by xster** [comment]
specifically re tests:

- check that the darkTheme is used when a mock window or the platform sends Brightness.dark
- check that the theme and darkTheme toggles automatically via a descendent Material widget when the platform toggles
- check that the theme is indeed used when darkTheme is undefined. 

In general, I think it's a good idea to run `flutter test --coverage` and making sure all the new lines are covered until we have an infra to display it in GitHub. You can see the lcov with a vscode extension or something (or you can just read the output file).

File: packages/flutter/lib/src/material/app.dart [hunk 11]
```diff
@@ -163,9 +166,41 @@ class MaterialApp extends StatefulWidget {
   /// This value is passed unmodified to [WidgetsApp.onGenerateTitle].
   final GenerateAppTitle onGenerateTitle;
 
-  /// The colors to use for the application's widgets.
+  /// The [ThemeData] to use when the platform's requested brightness is [Brightness.light] or undefined.
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: We try and keep doc comment lines to < 80 characters.

Although it's helpful to expand on the original one-line of information about this property, it seems to have become a little too focused its relationship to darkTheme etc.

How about this:
```
Default visual properties, like colors fonts and shapes, for this app's material widgets.

A second [darkTheme] [ThemeData] value, which is used when the underlying platform requests a "dark mode" UI, can also be specified.

The default value of this property is [ThemeData.light()].
```

File: packages/flutter/lib/src/material/app.dart [hunk 12]
```diff
@@ -163,9 +166,41 @@ class MaterialApp extends StatefulWidget {
...
+  /// that's also [Brightness.dark].
+  ///
+  /// Uses [theme] instead when null. Defaults to a light [ThemeData.fallback] theme when both [darkTheme] and
+  /// [theme] are null.
+  ///
+  /// See also:
+  ///
+  /// * [MediaQueryData.platformBrightness], which indicates the platform's desired brightness and is used to
+  ///    automatically toggle between [theme] and [darkTheme] in [MaterialApp].
+  /// * [ThemeData.brightness], which indicates the [Brightness] of a [ThemeData] you may want to use as a
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: This would be enough to say here:
```
[ThemeData.brightness], which indicates the [Brightness]  of a theme's colors
```

File: packages/flutter/lib/src/material/app.dart [hunk 13]
```diff
@@ -403,45 +438,63 @@ class _MaterialAppState extends State<MaterialApp> {
...
+      initialRoute: widget.initialRoute,
+      onGenerateRoute: widget.onGenerateRoute,
+      onUnknownRoute: widget.onUnknownRoute,
+      builder: (BuildContext context, Widget child) {
+        // Use a light theme, dark theme, or fallback theme.
+        ThemeData theme;
+        final ui.Brightness platformBrightness = MediaQuery.platformBrightnessOf(context);
+        if (platformBrightness == ui.Brightness.dark && widget.darkTheme != null) {
+          theme = widget.darkTheme;
+        } else if (widget.theme != null) {
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: or just
```
else
  theme = widget.theme ?? ThemeData.fallback();
```

File: packages/flutter/lib/src/rendering/binding.dart [hunk 14]
```diff
@@ -163,6 +164,12 @@ mixin RendererBinding on BindingBase, ServicesBinding, SchedulerBinding, Semanti
   /// See [Window.onTextScaleFactorChanged].
   @protected
   void handleTextScaleFactorChanged() { }
+  
+  /// Called when the platform brightness changes.
+  ///
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/binding.dart]: The would be a good place to explain where a handler would find the platform brightness value (I assume we want a statement about MediaQueryData here).

File: packages/flutter/lib/src/widgets/binding.dart [hunk 15]
```diff
@@ -214,6 +214,20 @@ abstract class WidgetsBindingObserver {
...
+  /// For example, starting in Android Pie, battery saver mode asks all apps to
+  /// render in a "dark mode". That change is reflected in this property.
+  ///
+  /// Not all platforms necessarily support a concept of brightness mode. Those
+  /// platforms will report [Brightness.light] in this property.
+  /// 
+  /// See also:
+  /// 
+  ///  * [MediaQuery.of], which provides a similar service with less
+  ///    boilerplate.
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/binding.dart]: Instead of "less boilerplate" you could mention that MediaQuery is an inherited widget whose dependents are automatically rebuilt..

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 16]
```diff
@@ -109,6 +110,15 @@ class MediaQueryData {
...
   final double textScaleFactor;
+  
+  /// The current brightness mode of the host platform.
+  /// 
+  /// For example, starting in Android Pie, battery saver mode asks all apps to
+  /// render in a "dark mode". That change is reflected in this property.
+  /// 
+  /// Not all platforms necessarily support a concept of brightness mode. Those
+  /// platforms will report [Brightness.light] in this property.
+  final ui.Brightness platformBrightness;
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/media_query.dart]: Is the `ui.` qualifier needed?

File: packages/flutter/test/widgets/media_query_test.dart [hunk 17]
```diff
@@ -47,6 +47,7 @@ void main() {
     expect(data.invertColors, false);
     expect(data.disableAnimations, false);
     expect(data.boldText, false);
+    expect(data.platformBrightness, ui.Brightness.light);
```
Comment (HansMuller) [inline packages/flutter/test/widgets/media_query_test.dart]: As before: I think Brightness should be available to all Flutter libraries without qualification.

**Review by HansMuller — COMMENTED** [review]
This looks good, just some small requests.

File: packages/flutter/lib/src/widgets/binding.dart [hunk 18]
```diff
@@ -214,6 +214,20 @@ abstract class WidgetsBindingObserver {
   ///    boilerplate.
   void didChangeTextScaleFactor() { }
 
+  /// Called when the platform's desired brightness changes.
+  ///
+  /// For example, starting in Android Pie, battery saver mode asks all apps to
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: I don't understand this and the next paragraph. It seems to be documenting a property, but it's on a method.

File: packages/flutter/lib/src/widgets/binding.dart [hunk 19]
```diff
@@ -214,6 +214,20 @@ abstract class WidgetsBindingObserver {
...
+  /// render in a "dark mode". That change is reflected in this property.
+  ///
+  /// Not all platforms necessarily support a concept of brightness mode. Those
+  /// platforms will report [Brightness.light] in this property.
+  /// 
+  /// See also:
+  /// 
+  ///  * [MediaQuery.of], which provides a similar service with less
+  ///    boilerplate.
+  void didChangePlatformBrightness() { }
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: It should probably be documented above how one could find out about the new platform brightness after this fires.

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 20]
```diff
@@ -1,7 +1,6 @@
 // Copyright 2015 The Chromium Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
-
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: nit: We usually do have one blank line after the license header, so maybe revert this change?

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 9 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/media_query.dart]: ACK

File: packages/flutter/lib/src/widgets/media_query.dart [hunk 21]
```diff
@@ -109,6 +110,15 @@ class MediaQueryData {
   ///  * [MediaQuery.textScaleFactorOf], a convenience method which returns the
   ///    textScaleFactor defined for a [BuildContext].
   final double textScaleFactor;
+  
+  /// The current brightness mode of the host platform.
+  /// 
+  /// For example, starting in Android Pie, battery saver mode asks all apps to
+  /// render in a "dark mode". That change is reflected in this property.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/media_query.dart]: This seems to just get you the current value? What do you mean by "change" here?

File: packages/flutter/test/widgets/media_query_test.dart [hunk 22]
```diff
@@ -225,6 +229,33 @@ void main() {
    expect(insideTextScaleFactor, 4.0);
  });
 
+  testWidgets('MediaQuery.platformBrightnessOf', (WidgetTester tester) async {
```
Comment (goderbauer) [inline packages/flutter/test/widgets/media_query_test.dart]: (I've not looked into how visible this is):

Any chance we can get a little more test coverage here? E.g. make sure that the appropriate callbacks are called when this value changes from the engine side.

File: packages/flutter/lib/src/material/app.dart (hunk 6 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/app.dart]: No change. I was just pointing out where the auto-rebuild inherited widget dependency was coming from.

File: packages/flutter/lib/src/material/app.dart [hunk 23]
```diff
@@ -163,9 +166,41 @@ class MaterialApp extends StatefulWidget {
   /// This value is passed unmodified to [WidgetsApp.onGenerateTitle].
   final GenerateAppTitle onGenerateTitle;
 
-  /// The colors to use for the application's widgets.
+  /// The [ThemeData] to use when the platform's requested brightness is [Brightness.light] or undefined.
+  ///
+  /// Host platforms such as Android Pie can request a system-wide "dark mode" when entering battery saver mode.
+  ///
+  /// When the host platform requests a [Brightness.light] mode, you may want to supply a [ThemeData.brightness]
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: I replaced these docs with a recommendation from Hans.  Can you check the docs again after I push the next change and see if you're OK with the new one?

File: packages/flutter/lib/src/widgets/binding.dart (hunk 18 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/binding.dart]: I replaced these docs as well based on a recommendation from Hans.  Please take another look after the next push.

File: packages/flutter/lib/src/widgets/binding.dart (hunk 19 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/binding.dart]: This information will be added in the next push.

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 20 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/media_query.dart]: Good catch.  Done.

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 21 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/media_query.dart]: I deleted the statement about "change"

File: packages/flutter/test/widgets/media_query_test.dart (hunk 22 above)
Comment (matthew-carroll) [reply packages/flutter/test/widgets/media_query_test.dart]: I'll look into additional tests.  Please take a look at the next push to see if the test coverage is acceptable.

File: packages/flutter/lib/src/material/app.dart (hunk 11 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: I changed it to your recommendation.

File: packages/flutter/lib/src/material/app.dart (hunk 12 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: I changed it to your recommendation.

File: packages/flutter/lib/src/material/app.dart (hunk 13 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Personally I prefer that if one conditional branch exists then all branches are explicit.

File: packages/flutter/lib/src/rendering/binding.dart (hunk 14 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/rendering/binding.dart]: I reworked the docs to show multiple query options.

File: packages/flutter/lib/src/widgets/binding.dart (hunk 15 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/binding.dart]: I changed it to your recommendation.

File: packages/flutter/lib/src/widgets/media_query.dart (hunk 16 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/media_query.dart]: I changed this file's import to bring in Brightness without a qualifier.

File: packages/flutter/test/widgets/media_query_test.dart (hunk 17 above)
Comment (matthew-carroll) [reply packages/flutter/test/widgets/media_query_test.dart]: I changed this file's import to bring in Brightness without a qualifier.

**Comment by matthew-carroll** [comment]
@HansMuller @goderbauer @xster - I just pushed a number of updates based on comments.  Can you all do another pass?

File: packages/flutter/lib/src/widgets/binding.dart [hunk 24]
```diff
@@ -213,6 +213,35 @@ abstract class WidgetsBindingObserver {
   ///    boilerplate.
   void didChangeTextScaleFactor() { }
 
+  /// Called when the platform brightness changes.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/binding.dart]: Looks like this comment is the same as above? Maybe use an `{@template}`.

**Review by goderbauer — APPROVED** [review]
LGTM if it looks good to @xster and @HansMuller as well.

**Comment by goderbauer** [comment]
Looks like cirrus is not too happy, though.

File: packages/flutter/test/material/app_test.dart [hunk 25]
```diff
@@ -435,4 +435,96 @@ void main() {
...
+    );
+
+    // Switch the platformBrightness from light to dark and pump the widget tree
+    // to process changes.
+    binding.window.platformBrightnessTestValue = Brightness.dark;
+    await tester.pumpAndSettle();
+
+    expect(themeBeforeBrightnessChange.brightness, Brightness.light);
+    expect(themeAfterBrightnessChange.brightness, Brightness.dark);
+  });
```
Comment (xster) [inline packages/flutter/test/material/app_test.dart]: Can you add more fallback behavior tests? e.g. light is used when window is dark and darkTheme is not supplied, light theme is not supplied, changing from a darkTheme to a light window still works because it fallbacks to ThemeData.fallback etc?

**Review by xster — COMMENTED** [review]
LGTM. Thanks for adding tests.

File: packages/flutter/lib/src/widgets/binding.dart (hunk 24 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/widgets/binding.dart]: Done

File: packages/flutter/lib/src/material/app.dart [hunk 26]
```diff
@@ -163,9 +166,45 @@ class MaterialApp extends StatefulWidget {
...
   final GenerateAppTitle onGenerateTitle;
 
-  /// The colors to use for the application's widgets.
+  /// Default visual properties, like colors fonts and shapes, for this app's
+  /// material widgets.
+  ///
+  /// A second [darkTheme] [ThemeData] value, which is used when the underlying
+  /// platform requests a "dark mode" UI, can also be specified.
+  ///
+  /// The default value of this property is [ThemeData.light()].
```
[the remark this answers is not in the document]
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: I just copied exactly what Hans wrote, but I've adjusted that reference to be ThemeData.fallback

File: packages/flutter/test/material/app_test.dart (hunk 25 above)
Comment (matthew-carroll) [reply packages/flutter/test/material/app_test.dart]: Done

File: packages/flutter/lib/src/material/app.dart [hunk 27]
```diff
@@ -163,9 +166,45 @@ class MaterialApp extends StatefulWidget {
...
   final GenerateAppTitle onGenerateTitle;
 
-  /// The colors to use for the application's widgets.
+  /// Default visual properties, like colors fonts and shapes, for this app's
+  /// material widgets.
+  ///
+  /// A second [darkTheme] [ThemeData] value, which is used when the underlying
+  /// platform requests a "dark mode" UI, can also be specified.
+  ///
+  /// The default value of this property is a light [ThemeData.fallback].
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: This doesn't read very well. I think you mean: a light theme, the value of [ThemeData.fallback]. However [TheemData.fallback] doesn't seem like a useful value to mention, since it's hardwired to ThemeData.light().

So: The default value of this property is the value of [ThemeData.light()]. That's equally true and easier to understand.

File: packages/flutter/lib/src/material/app.dart [hunk 28]
```diff
@@ -163,9 +166,45 @@ class MaterialApp extends StatefulWidget {
...
+  /// The [ThemeData] to use when the platform specifically requests a dark
+  /// themed UI.
+  ///
+  /// Host platforms such as Android Pie can request a system-wide "dark mode"
+  /// when entering battery saver mode.
+  ///
+  /// When the host platform requests a [Brightness.dark] mode, you may want to
+  /// supply a [ThemeData.brightness] that's also [Brightness.dark].
+  ///
+  /// Uses [theme] instead when null. Defaults to a light [ThemeData.fallback]
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: Inserting ThemeData.fallback into a sentence here has the same problem mentioned earlier.

File: packages/flutter/lib/src/material/app.dart [hunk 29]
```diff
@@ -163,9 +166,45 @@ class MaterialApp extends StatefulWidget {
...
+  /// theme when both [darkTheme] and [theme] are null.
+  ///
+  /// See also:
+  ///
+  /// * [MediaQueryData.platformBrightness], which indicates the platform's
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: See also formatting is:
```
///  * foo...
///    bar
```

Two spaces after the `///`; four spaces after the `///` for the second and subsequent lines.

File: packages/flutter/lib/src/material/app.dart [hunk 30]
```diff
@@ -163,9 +166,45 @@ class MaterialApp extends StatefulWidget {
...
+  ///    desired brightness and is used to automatically toggle between [theme]
+  ///    and [darkTheme] in [MaterialApp].
+  /// * [ThemeData.brightness], which indicates the [Brightness] of a
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: I think it would be simpler to just say:
which is typically set to the value of [MediaQueryData.platformBrightness].

File: packages/flutter/lib/src/material/app.dart [hunk 31]
```diff
@@ -403,45 +442,80 @@ class _MaterialAppState extends State<MaterialApp> {
...
+          theme = ThemeData.fallback();
+        }
+
+        return AnimatedTheme(
+          data: theme,
+          isMaterialAppTheme: true,
+          child: widget.builder != null
+              ? Builder(
+                  builder: (BuildContext context) {
+                    // Why are we surrounding a builder with a builder?
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/app.dart]: This seems like a generic problem with AnimatedTheme. Maybe that's where the extra BuildContext should be inserted. 

Could be fixed in a separate PR.

**Review by HansMuller — APPROVED** [review]
LGTM with a few suggestions

File: packages/flutter/lib/src/material/app.dart (hunk 27 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Done

File: packages/flutter/lib/src/material/app.dart (hunk 28 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Changed verbiage to match your other comment.

File: packages/flutter/lib/src/material/app.dart (hunk 29 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Added a space to each "*" line in these two docs.

File: packages/flutter/lib/src/material/app.dart (hunk 30 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Done

File: packages/flutter/lib/src/material/app.dart (hunk 31 above)
Comment (matthew-carroll) [reply packages/flutter/lib/src/material/app.dart]: Yeah, I'll let that be investigated on its own.  Seems like it could be a deep investigation with regard to performance implications and also backwards compatibility.

**Comment by creativecreatorormaybenot** [comment]
@matthew-carroll How can I change the `platformBrightness` value and also affect the `MaterialApp` theme? I implemented adding a `MediaQuery` with my custom value for `platformBrightness` similar to [how it was done here](https://stackoverflow.com/a/52012115/6509751). I even tried setting state of the widget containing the `MaterialApp` on via `WidgetBindingsObserver` `didChangePlatformBrightness`. The `MediaQuery` has the correct value down the tree, but the theme does not change accordingly. However, when I toggle the device to dark mode it works fine. What part am I missing?

**Comment by matthew-carroll** [comment]
@creativecreatorormaybenot can you please file this question as an issue instead of a PR comment? Also, can you please include as much of your actual code as possible?  I need to see exactly what you're doing to be able to comment.
